### PR TITLE
bfm/bfm_cobra3.cpp: attempt to improve non-MPEG hardware parts.

### PR DESCRIPTION
**Pull request by james-wallace-ghub**
New systems marked not working
------------------------------
Telly Addicts (New Series) (Bellfruit) (Cobra 3) [James Wallace, blueonesarefaster]

bfm_cobra3.cpp - hook up revised CD dumps and non-MPEG graphics hardware [James Wallace, blueonesarefaster]

### CODE REVIEW

File: src/mame/bfm/bfm_cobra3.cpp [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+Music TR9C1710 (Video RAMDAC)
+
+Part number :   STV8438CV
+ Manufacture :  ST Microelectronics
+Specification(s)
+ Description :   8-Bit Triple DAC with Analog Switch
+ Package :   PQFP44
+
+MC68340 : Integrated Processor with DMA
+ The MC68340 is a high-performance 32-bit integrated processor with direct memory access (DMA), combining an enhanced M68000-compatible processor, 32-bit DMA, and other peripheral subsystems on a single integrated circuit. The MC68340 CPU32 delivers 32-bit CISC processor performance from a lower cost 16-bit memory system. The combination of peripherals offered in the MC68340 can be found in a diverse range of microprocessor-based systems. Systems requiring very high-speed block transfers of data can benefit from the MC68340.
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: This sounds like a direct quote from some spec-sheet or leaflet. Is it really necessary to have it?

File: src/mame/bfm/bfm_cobra3.cpp [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+l5380 -scsi controller
+
+ymz280b - in MAME - sound chip
+
+audio board
+
+75585116 - PALCE 16V8H
+75585118 - PALCE 16V8H
+
+TI introduce the TMS320AV110 MPEG audio decoder based on TI's 16 bits
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: Based on TI's 16 bits what, exactly? This sentence seems to start and then

File: src/mame/bfm/bfm_cobra3.cpp [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+Four Chip Selects
+System Integration
+32 Address Lines, 16 Data Lines
+Low Power Management
+4.8 MIPS @ 25 MHz
+Available in 16 and 25 MHz
+Available in 3.3 and 5V
+
+
+l5380 -scsi controller
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: Why is there inconsistent capitalization here and following, compared to the proper capitalization above? This just indicates that the above is copy/pasted directly from some spec-sheet, which we do not do due to copyright.

File: src/mame/bfm/bfm_cobra3.cpp [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+			return m_cpuregion[offset & 0x7ffff];
+		case 2://RAM
+			return m_mainram[offset & 0x1fff];
+
+		case 3: //I/O 00ac0000-00ac07ff
+			{
+				offset &= 0x7ff;
+				offs_t cs_addr_8_11 = (offset * 2) & 0xf00;
+
+				switch(cs_addr_8_11)
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp:133]: Please ensure consistent spacing, per https://docs.mamedev.org/contributing/cxx.html#spacing (compare to line 139).

File: src/mame/bfm/bfm_cobra3.cpp [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+				{
+					case 0x300: //ac0300
+						if(ACCESSING_BITS_0_7)
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp:136]: Ibid, and through the rest of the file.

File: src/mame/bfm/bfm_cobra3.cpp [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+						logerror("%08x maincpu read access offset %08x mem_mask %08x cs %d\n", pc, offset*4, mem_mask, cs);
+						break;
+				}
+			}
+			break;
+
+		case 4: //0x00b10000 SCSI controller
+				if(ACCESSING_BITS_8_15)
+				{
+					return(m_scsic->read(offset & 0x0f)<<8);
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: Ibid. Also, please fix your tabs.

File: src/mame/bfm/bfm_cobra3.cpp [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+				}
+				break;
+
+		default:
+				logerror("%08x maincpu read access offset %08x mem_mask %08x cs %d\n", pc, offset*4, mem_mask, cs);
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: Tabbing, again.

File: src/mame/bfm/bfm_cobra3.cpp [hunk 8]
```diff
@@ -0,0 +1,592 @@
...
+
+void bfm_cobra3_state::bfm_cobra3_mem_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	int pc = m_maincpu->pc();
+	int cs = m_maincpu->get_cs(offset * 2);
+
+	switch (cs)
+	{
+		case 1://ROM, shouldn't write here?
+				logerror("%08x maincpu write access(1) offset %08x data %08x mem_mask %08x cs %d\n", pc, offset*4, data, mem_mask, cs);
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: As above, and throughout the file.

Also, please use `machine().describe_context()`, not m_maincpu->pc().

File: src/mame/bfm/bfm_cobra3.cpp [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+						break;
+
+					case 0x300:
+						if(ACCESSING_BITS_0_7)
+						{
+							m_ymz->write(offset & 1, data);
+						}
+						break;
+
+					case 0x500://SCSI DMA
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: Spacing, again. Please try to maintain a space between the comment marker and the comment text.

File: src/mame/bfm/bfm_cobra3.cpp [hunk 10]
```diff
@@ -0,0 +1,592 @@
...
+
+					default:
+						// coin divert, hoppers, note validator
+						logerror("%08x maincpu write access(3) offset %08x data %08x mem_mask %08x cs %d\n", pc, offset*4, data, mem_mask, cs);
+						break;
+				}
+			}
+			break;
+			
+		case 4: //0x00b10000 SCSI controller
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: "0x00b10000"?

File: src/mame/bfm/bfm_cobra3.cpp [hunk 11]
```diff
@@ -0,0 +1,592 @@
...
+
+static void cobra_scsi_devices(device_slot_interface &device)
+{
+	device.option_add("cdrom", NSCSI_CDROM);
+	device.option_add_internal("ncr53c80", NCR53C80);
+}
+
+void bfm_cobra3_state::dma1_drq(int state)
+{
+	//? Can't see this
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: ? Can't understand this

File: src/mame/bfm/bfm_cobra3.cpp [hunk 12]
```diff
@@ -0,0 +1,592 @@
...
+	m_maincpu->set_input_line(5, !state);
+}
+
+uint32_t bfm_cobra3_state::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	if(m_scc66470->display_enabled())
+	{
+		if(cliprect.min_y == cliprect.max_y)
+		{
+			uint32_t *dest;
```
Comment (MooglyGuy) [inline src/mame/bfm/bfm_cobra3.cpp]: No need for C89-style variable declaration anymore.

**Review by MooglyGuy — DISMISSED** [review]

File: src/mame/bfm/bfm_cobra3.cpp (hunk 1 above)
Comment (james-wallace-ghub) [reply src/mame/bfm/bfm_cobra3.cpp]: Will simplify this to a straight TODO.

File: src/mame/bfm/bfm_cobra3.cpp (hunk 2 above)
Comment (james-wallace-ghub) [reply src/mame/bfm/bfm_cobra3.cpp]: Ditched, see above. What I meant was the TI '16 bit' interface, as opposed to the 8 bit. It harms more than it helps anyway, and given the CPU stuff I could be wrong anyway. When in doubt, take it out.

File: src/mame/bfm/bfm_cobra3.cpp (hunk 12 above)
Comment (james-wallace-ghub) [reply src/mame/bfm/bfm_cobra3.cpp]: So, I can do a one liner for definitions? OK, will do.

File: src/mame/bfm/bfm_cobra3.cpp (hunk 10 above)
Comment (james-wallace-ghub) [reply src/mame/bfm/bfm_cobra3.cpp]: This was me trying to rationalise CS access to an old school memory map for rationalisation. Not really needed, so I've removed this and other similar lines.

File: src/devices/machine/68340.cpp [hunk 13]
```diff
@@ -130,8 +130,16 @@ uint8_t m68340_cpu_device::int_ack(offs_t offset)
...
-	return ((!BIT(offset, 0) ? (m_m68340_base >> 16): m_m68340_base)) & 0xffff;
+		LOGMASKED(LOG_BASE, "%08x m68340_internal_base_r %08x, (%08x) (%08x)\n", m_ppc, offset*2,mem_mask, m_sfc);
+
+	if (m_sfc==0x7)
+	{
+		return ((!BIT(offset, 0) ? (m_m68340_base >> 16): m_m68340_base)) & 0xffff;
+	}
+	else
+	{
+		return -1;
```
Comment (angelosa) [inline src/devices/machine/68340.cpp]: `space.unmap()`

Comment (james-wallace-ghub) [reply src/devices/machine/68340.cpp]: Addressed, but needs unmap_high, as the internal test expects unmapped values to return non zero.

File: src/mame/bfm/bfm_cobra3.cpp [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+			m_scsibus(*this, "scsi"),
+			m_scsic(*this, "scsi:6:ncr5380"),
+			m_watchdog(*this, "watchdog")
+	{ }
+
+	std::unique_ptr<uint16_t[]> m_mainram;
+
+	void volume_control(uint8_t direction, uint8_t clock);
+	uint16_t bfm_cobra3_mem_r(offs_t offset, uint16_t mem_mask = ~0);
+	void bfm_cobra3_mem_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
```
Comment (angelosa) [inline src/mame/bfm/bfm_cobra3.cpp]: Please reduce this naming, `mem_r` / `mem_w` should suffice.

File: src/mame/bfm/bfm_cobra3.cpp [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+				}
+			}
+			if(*src == 254)
+			{
+				// finally 32 pixels of mpeg border colour when it's mixed in
+			}
+			else
+			{
+				std::fill_n(dest, 32, m_palette->pen(*src));
+			}
```
Comment (angelosa) [inline src/mame/bfm/bfm_cobra3.cpp]: I think you can reduce this to just do `if (*src != 254)` unless you eventually need the branch where you have the comment in (and that would be a `TODO`)

Comment (james-wallace-ghub) [reply src/mame/bfm/bfm_cobra3.cpp]: Revised the comment

File: src/mame/bfm/bfm_cobra3.cpp (hunk 14 above)
Comment (james-wallace-ghub) [reply src/mame/bfm/bfm_cobra3.cpp]: Agreed, did the cleanup accordingly.
